crates/qasm3/src/circuit.rs

```diff
@@ -0,0 +1,333 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2024
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use pyo3::prelude::*;
+use pyo3::types::{PyList, PyString, PyTuple, PyType};
+
+use crate::error::QASM3ImporterError;
+
+pub trait PyRegister {
+    fn bit(&self, py: Python, index: usize) -> PyResult<Py<PyAny>>;
+    // This really should be
+    //      fn iter<'a>(&'a self, py: Python<'a>) -> impl Iterator<Item = &'a PyAny>;
+    // or at a minimum
+    //      fn iter<'a>(&'a self, py: Python<'a>) -> ::pyo3::types::iter::PyListIterator<'a>;
+    // but we can't use the former before Rust 1.75 and the latter before PyO3 0.21.
+    fn bit_list<'a>(&'a self, py: Python<'a>) -> &'a PyList;
+}
+
+macro_rules! register_type {
+    ($name: ident) => {
+        /// Rust-space wrapper around Qiskit `Register` objects.
+        pub struct $name {
+            /// The actual register instance.
+            object: Py<PyAny>,
+            /// A pointer to the inner list of bits.  We keep a handle to this for lookup
+            /// efficiency; we can use direct list methods to retrieve the bit instances, rather
+            /// than needing to indirect through the general `__getitem__` of the register, or
+            /// looking up the qubit instances on the circuit.
+            items: Py<PyList>,
+        }
+
+        impl PyRegister for $name {
+            /// Get an individual bit from the register.
+            fn bit(&self, py: Python, index: usize) -> PyResult<Py<PyAny>> {
+                // Unfortunately, `PyList::get_item_unchecked` isn't usable with the stable ABI.
+                self.items
+                    .as_ref(py)
+                    .get_item(index)
+                    .map(|item| item.into_py(py))
+            }
+
+            fn bit_list<'a>(&'a self, py: Python<'a>) -> &'a PyList {
+                self.items.as_ref(py)
+            }
+        }
+
+        impl ::pyo3::IntoPy<Py<PyAny>> for $name {
+            fn into_py(self, _py: Python) -> Py<PyAny> {
+                self.object
+            }
+        }
+
+        impl ::pyo3::ToPyObject for $name {
+            fn to_object(&self, _py: Python) -> Py<PyAny> {
+                // _Technically_, allowing access this internal object can let the Rust-space
+                // wrapper get out-of-sync since we keep a direct handle to the list, but in
+                // practice, the field it's viewing is private and "inaccessible" from Python.
+                self.object.clone()
```

Comment:
```suggestion
            fn to_object(&self, py: Python) -> Py<PyAny> {
                // _Technically_, allowing access this internal object can let the Rust-space
                // wrapper get out-of-sync since we keep a direct handle to the list, but in
                // practice, the field it's viewing is private and "inaccessible" from Python.
                self.object.clone_ref(py)
```

Reply:
Ha yeah - iirc I wrote this, then about 5 minutes later realised that `Py::clone_ref` exists and that I should have been using it everywhere.  Not surprised I missed a couple of places.

Reply:
Done in 3345c34f22.